fixup MUNCH for BSPs using motorola_powerpc bootloader

mdavidsaver opened the pull request:
Alternative to #217

For RTEMS-mvme2700, extract the bootloader+rtems.gz as "binary".  Also fix both mvme2100 and 2700 to avoid parallel linking issue w/ "rtems.gz" filename (hard-coded in bootloader linker script).

mdavidsaver committed:
fixup MUNCH for BSPs using motorola_powerpc bootloader

mdavidsaver commented:
@mrippa, maybe simpler to show what I am thinking.

mdavidsaver commented:
For reference, the bootloader source https://github.com/RTEMS/rtems/tree/master/bsps/powerpc/motorola_powerpc/bootloader

mdavidsaver commented:
Maybe worth noting that the mvme3100 BSP uses a [different scheme](https://github.com/RTEMS/rtems/blob/master/bsps/powerpc/shared/start/preload.S) for the same task, where the chain bootloader is part of the user executable.

anjohnson commented:
I haven't tested this at all, but using the subdirectory to avoid having to modify ppcboot.lds seems like an excellent idea, thanks! Hopefully it fixes @mrippa's issue too.

mrippa commented:
The new rule adds a necessary change to the ppcboot.lds. Testing this the location of rtems.gz also needs to include the directory libComTestHarness.dir/rtems.gz(*). I think a modified file from rtems-devs could suffice, but I wanted to test.

I modified a local copy of ppcboot.lds and tested it. 

```diff
--- /gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/ppcboot.lds   2021-09-16 18:47:42.000000000 +0000
+++ /home/mrippa/ppcboot.lds    2022-02-07 22:03:14.642121629 +0000
@@ -47,7 +47,7 @@
 /*  . = ALIGN(16); */
   .image :
   {
-    rtems.gz(*)
+    libComTestHarness.dir/rtems.gz(*)
     . = ALIGN(4);
     *.gz(*)
   }
```

This results in the some symbols not being defined:
```sh
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-gcc --pipe -B/gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/ -specs bsp_specs -qrtems   -fasm   -mcpu=604 -mmultiple -mstring -mstrict-align -meabi       -D_GNU_SOURCE -D_DEFAULT_SOURCE            -DUNIX      -O2 -g -ffunction-sections -fdata-sections   -Wall -Werror-implicit-function-declaration       -DMY_DO_BOOTP=NULL -DHAVE_PPCBUG -DNVRAM_INDIRECT    -D__LINUX_ERRNO_EXTENSIONS__ -DRTEMS_LEGACY_STACK  -I. -I../O.Common -I. -I. -I.. -I../../../../include/compiler/gcc -I../../../../include/os/RTEMS -I../../../../include         -c rtemsTestData.c
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-g++ -B/gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/ -specs bsp_specs -qrtems   -o libComTestHarness -static -L/home/mrippa/work/vendor/epics-base/lib/RTEMS-mvme2700           -mcpu=604 -mmultiple -mstring -mstrict-align -meabi -u POSIX_Init         epicsTypesTest.o epicsInlineTest1.o epicsInlineTest2.o epicsInlineTest3.o epicsInlineTest4.o epicsCalcTest.o epicsAlgorithmTest.o epicsMathTest.o epicsMMIOTest.o epicsEllTest.o epicsEnvTest.o epicsEnvUnsetTest.o epicsErrlogTest.o epicsStdioTest.o epicsStdlibTest.o epicsSockResolveTest.o epicsStringTest.o epicsTimeTest.o epicsThreadTest.o epicsThreadClassTest.o epicsThreadOnceTest.o epicsThreadPriorityTest.o epicsThreadPrivateTest.o epicsThreadHooksTest.o epicsThreadPoolTest.o initHookTest.o epicsExitTest.o epicsTimerTest.o ringPointerTest.o ringBytesTest.o epicsEventTest.o epicsMutexTest.o epicsSpinTest.o epicsAtomicTest.o macDefExpandTest.o cvtFastTest.o macLibTest.o aslibtest.o taskwdTest.o blockingSockTest.o epicsMessageQueueTest.o epicsStackTraceTest.o ipAddrToAsciiTest.o osiSockTest.o epicsRunLibComTests.o epicsThreadPerform.o epicsMaxThreads.o buckTest.o epicsAtomicPerform.o cvtFastPerform.o epicsTimeZoneTest.o rtemsTestHarness.o rtemsTestData.o   -lCom       -lm -lrtemsCom -lCom -ltftpfs -lnfs -lz -ltelnetd -lrtemscpu -lc -lm -lgcc
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-objcopy -O binary -R .comment -S libComTestHarness libComTestHarness.dir/rtems
gzip -f9 libComTestHarness.dir/rtems
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-ld -o libComTestHarness.boot.elf /gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/bootloader.o --just-symbols=libComTestHarness -b binary libComTestHarness.dir/rtems.gz --no-warn-mismatch -T /home/mrippa/ppcboot.lds -Map libComTestHarness.boot.map
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-ld:/home/mrippa/ppcboot.lds:96: undefined symbol `_binary_rtems_gz_end' referenced in expression
make[4]: *** [../../../../configure/RULES_BUILD:353: libComTestHarness.boot] Error 1
make[4]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules/libcom/test/O.RTEMS-mvme2700'
make[3]: *** [../../../configure/RULES_ARCHS:58: build.RTEMS-mvme2700] Error 2
make[3]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules/libcom/test'
make[2]: *** [../../configure/RULES_DIRS:85: test.build.RTEMS-mvme2700] Error 2
make[2]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules/libcom'
make[1]: *** [../configure/RULES_DIRS:85: libcom.build.RTEMS-mvme2700] Error 2
make[1]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules'
make: *** [configure/RULES_DIRS:85: modules.build.RTEMS-mvme2700] Error 2
```

For reference here's the modified ppcboot.lds
```
OUTPUT_ARCH(powerpc)
OUTPUT_FORMAT ("elf32-powerpc", "elf32-powerpc", "elf32-powerpc")
/* Do we need any of these for elf?
   __DYNAMIC = 0;    */
SECTIONS
{
  .text :
  {
    /* We have to build the header by hand, painful since ppcboot
        format support is very poor in binutils. 
        objdump -b ppcboot zImage --all-headers can be used to check. */
    /* The following line can be added as a branch to use the same image
     * for netboot as for prepboots, the only problem is that objdump
     * did not in this case recognize the format since it insisted
     * in checking the x86 code area held only zeroes. 
     */
    LONG(0x48000000+start);
    . = 0x1be; BYTE(0x80); BYTE(0)
    BYTE(2); BYTE(0); BYTE(0x41); BYTE(1);
    BYTE(0x12); BYTE(0x4f); LONG(0);
    BYTE(((_edata + 0x1ff)>>9)&0xff); 
    BYTE(((_edata + 0x1ff)>>17)&0xff); 
    BYTE(((_edata + 0x1ff)>>25)&0xff);
    . = 0x1fe;
    BYTE(0x55);
    BYTE(0xaa);
    BYTE(start&0xff);
    BYTE((start>>8)&0xff);
    BYTE((start>>16)&0xff);
    BYTE((start>>24)&0xff);
    BYTE(_edata&0xff);
    BYTE((_edata>>8)&0xff);
    BYTE((_edata>>16)&0xff);
    BYTE((_edata>>24)&0xff);
    BYTE(0); /* flags */
    BYTE(0); /* os_id */
    BYTE(0x4C); BYTE(0x69); BYTE(0x6e); 
    BYTE(0x75); BYTE(0x78); /* Partition name */
    . = 0x400;
    *(.text)
    *(.text*)
    *(.sdata2)
    *(.sdata2*)
    *(.rodata)
    *(.rodata*)
  }
/*  . = ALIGN(16); */
  .image :
  {
    libComTestHarness.dir/rtems.gz(*)
    . = ALIGN(4);
    *.gz(*)
  }
  /* Read-write section, merged into data segment: */
  /* . = ALIGN(4096); */
  .reloc   :
  {
    *(.got)
    _GOT2_TABLE_ = .;
    *(.got2)
    _FIXUP_TABLE_ = .;
    *(.fixup)
    _FIXUP_END_ = .;
  }

  .handlers : 
  {
    *(.exception)
  }

  .data    :
  {
    *(.data)
    *(.data*)
    *(.sdata)
    *(.sdata*)
    . = ALIGN(4);
    _data_end = .;
  }
  .bss :
  {
    *(.sbss)
    *(.sbss*)
    *(.bss)
    *(.bss*)
    *(COMMON)
    . = ALIGN(4);
    _bss_end = .;
  }
  .abs 0 : {
    __got2_entries = ABSOLUTE((_FIXUP_TABLE_ - _GOT2_TABLE_) >>2);
    __fixup_entries = ABSOLUTE((_FIXUP_END_ - _FIXUP_TABLE_)>>2);
    _edata = ABSOLUTE(_data_end);
    PROVIDE(_binary_initrd_gz_start = ABSOLUTE(0));
    PROVIDE(_binary_initrd_gz_end = ABSOLUTE(0));
    _rtems_gz_size = ABSOLUTE(_binary_rtems_gz_end - _binary_rtems_gz_start);
    _rtems_size = ABSOLUTE(__rtems_end - __rtems_start);
    __bss_words = ABSOLUTE(SIZEOF(.bss)>>2);
    __size = ABSOLUTE(_bss_end);
  }

  /DISCARD/ : 
  {
    *(.comment*)
    *(.debug*)
  }
}
```

mdavidsaver commented:
> The new rule adds a necessary change to the ppcboot.lds.

This is exactly what I was hoping to avoid.  Requiring users to manually patch each RTEMS install is imo. a last resort.

Interesting that `ld` doesn't complain about not being able to find `rtems.gz`...

mdavidsaver committed:
further fixup MUNCH for BSPs using motorola_powerpc bootloader

Avoid depending on rtems.gz name entirely by including contents
via. the ".incbin" assembler directive.

mdavidsaver commented:
If the `rtems.gz` file need not exist, then the real requirement is limited to the symbols `_binary_rtems_gz_start` and `_binary_rtems_gz_end`, which can be provided by other means.  Since we're only dealing with binutils, I think the easiest way to do this is the `.incbin` assembler directive.

@mrippa, please update and retest.

@anjohnson If this works, you'll want to look at the Makefile changes.  In particular the `ppc_bug.S` filename, and lack of quoting with `-IRTEMS_GZ=$*.app.bin.gz`.

mrippa commented:
I think there may be an issue with the vpath. I didn't have a lot of time today. I'll try on Wednesday.

```
make[1]: Entering directory '/home/mrippa/work/vendor/epics-base/modules/libcom/test/O.RTEMS-mvme2700'
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-objcopy -O binary -R .comment -S libComTestHarness libComTestHarness.app.bin
gzip -f9 libComTestHarness.app.bin
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-gcc --pipe -B/gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/ -specs bsp_specs -qrtems   -fasm   -mcpu=604 -mmultiple -mstring -mstrict-align -meabi       -D_GNU_SOURCE -D_DEFAULT_SOURCE            -DUNIX      -O2 -g -ffunction-sections -fdata-sections   -Wall -Werror-implicit-function-declaration       -DMY_DO_BOOTP=NULL -DHAVE_PPCBUG -DNVRAM_INDIRECT    -D__LINUX_ERRNO_EXTENSIONS__ -DRTEMS_LEGACY_STACK  -I. -I../O.Common -I. -I. -I.. -I../../../../include/compiler/gcc -I../../../../include/os/RTEMS -I../../../../include         -IRTEMS_GZ=libComTestHarness.app.bin.gz -c ppc_bug.S -o libComTestHarness.boot.o
powerpc-rtems5-gcc: error: ppc_bug.S: No such file or directory
powerpc-rtems5-gcc: fatal error: no input files
compilation terminated.
make[1]: *** [../../../../configure/RULES_BUILD:353: libComTestHarness.boot] Error 1
make[1]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules/libcom/test/O.RTEMS-mvme2700'
make: *** [../../../configure/RULES_ARCHS:58: install.RTEMS-mvme2700] Error 2
```

If I modify the compile to find ppc_bug.S it has trouble with the RTEMS_GZ define. 
```
[mrippa@e7r5-tr1 O.RTEMS-mvme2700]$ /gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-gcc --pipe -B/gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/ -specs bsp_specs -qrtems   -fasm   -mcpu=604 -mmultiple -mstring -mstrict-align -meabi       -D_GNU_SOURCE -D_DEFAULT_SOURCE            -DUNIX      -O2 -g -ffunction-sections -fdata-sections   -Wall -Werror-implicit-function-declaration       -DMY_DO_BOOTP=NULL -DHAVE_PPCBUG -DNVRAM_INDIRECT    -D__LINUX_ERRNO_EXTENSIONS__ -DRTEMS_LEGACY_STACK  -I. -I../O.Common -I. -I. -I.. -I../../../../include/compiler/gcc -I../../../../include/os/RTEMS -I../../../../include     -c ../../../../include/os/RTEMS/ppc_bug.S -o libComTestHarness.boot.o
../../../../include/os/RTEMS/ppc_bug.S: Assembler messages:
../../../../include/os/RTEMS/ppc_bug.S:23: Error: file not found: RTEMS_GZ
```

Replacing with   -DRTEMS_GZ... works better

```diff
@@ -21,7 +21,7 @@ MUNCH_SUFFIX = .boot
 define MUNCH_CMD
        $(RTEMS_TOOLS)/bin/$(OBJCOPY_FOR_TARGET) -O binary -R .comment -S $< $*.app.bin
        gzip -f9 $*.app.bin
-       $(COMPILE.c) -IRTEMS_GZ=$*.app.bin.gz -c ppc_bug.S -o $*.boot.o
+       $(COMPILE.c) -DRTEMS_GZ=$*.app.bin.gz -c ppc_bug.S -o $*.boot.o
        $(RTEMS_TOOLS)/bin/$(LD_FOR_TARGET) -o $@.elf \
                 $(PROJECT_RELEASE)/lib/bootloader.o \
                --just-symbols=$< \
```

mdavidsaver committed:
further fixup of fixup of MUNCH for BSPs using motorola_powerpc bootloader

mdavidsaver commented:
> Replacing with -DRTEMS_GZ... works better

Ah, yes.  I expect that it will...  oops

mdavidsaver committed:
ppc_bug.S use rodata

OS image blob should be read-only to speed up boot.

mrippa commented:
> Interesting that ld doesn't complain about not being able to find rtems.gz...

I missed that you changed things based on that. It always complained about finding rtems.gz and originally I modified ppc_bug.lds to include the path `libComTestHarness.dir/rtems.gz(*)`

I don't know why the ppc_bug.S path was not working with your compile line. The vpath looked right. I had to make this change:

```diff
[mrippa@e7r5-tr1 epics-base]$ git diff configure/os/CONFIG.Common.RTEMS-mvme2700
diff --git a/configure/os/CONFIG.Common.RTEMS-mvme2700 b/configure/os/CONFIG.Common.RTEMS-mvme2700
index 2f5a568ac..7d6ca1144 100644
--- a/configure/os/CONFIG.Common.RTEMS-mvme2700
+++ b/configure/os/CONFIG.Common.RTEMS-mvme2700
@@ -21,7 +21,9 @@ MUNCH_SUFFIX = .boot
 define MUNCH_CMD
        $(RTEMS_TOOLS)/bin/$(OBJCOPY_FOR_TARGET) -O binary -R .comment -S $< $*.app.bin
        gzip -f9 $*.app.bin
-       $(COMPILE.c) -DRTEMS_GZ=$*.app.bin.gz -c ppc_bug.S -o $*.boot.o
+       $(COMPILE.c) -DRTEMS_GZ=$*.app.bin.gz \
+                -c ../../../../include/os/RTEMS/ppc_bug.S \
+                -o $*.boot.o
        $(RTEMS_TOOLS)/bin/$(LD_FOR_TARGET) -o $@.elf \
                 $(PROJECT_RELEASE)/lib/bootloader.o \
                --just-symbols=$< \
```
This passes the compile and fails again on the link stage

```/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-g++ -B/gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/ -specs bsp_specs -qrtems   -o libComTestHarness -static -L/home/mrippa/work/vendor/epics-base/lib/RTEMS-mvme2700           -mcpu=604 -mmultiple -mstring -mstrict-align -meabi -u POSIX_Init         epicsTypesTest.o epicsInlineTest1.o epicsInlineTest2.o epicsInlineTest3.o epicsInlineTest4.o epicsCalcTest.o epicsAlgorithmTest.o epicsMathTest.o epicsMMIOTest.o epicsEllTest.o epicsEnvTest.o epicsEnvUnsetTest.o epicsErrlogTest.o epicsStdioTest.o epicsStdlibTest.o epicsSockResolveTest.o epicsStringTest.o epicsTimeTest.o epicsThreadTest.o epicsThreadClassTest.o epicsThreadOnceTest.o epicsThreadPriorityTest.o epicsThreadPrivateTest.o epicsThreadHooksTest.o epicsThreadPoolTest.o initHookTest.o epicsExitTest.o epicsTimerTest.o ringPointerTest.o ringBytesTest.o epicsEventTest.o epicsMutexTest.o epicsSpinTest.o epicsAtomicTest.o macDefExpandTest.o cvtFastTest.o macLibTest.o aslibtest.o taskwdTest.o blockingSockTest.o epicsMessageQueueTest.o epicsStackTraceTest.o ipAddrToAsciiTest.o osiSockTest.o epicsRunLibComTests.o epicsThreadPerform.o epicsMaxThreads.o buckTest.o epicsAtomicPerform.o cvtFastPerform.o epicsTimeZoneTest.o rtemsTestHarness.o rtemsTestData.o   -lCom       -lm -lrtemsCom -lCom -ltftpfs -lnfs -lz -ltelnetd -lrtemscpu -lc -lm -lgcc
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-objcopy -O binary -R .comment -S libComTestHarness libComTestHarness.app.bin
gzip -f9 libComTestHarness.app.bin
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-gcc --pipe -B/gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/ -specs bsp_specs -qrtems   -fasm   -mcpu=604 -mmultiple -mstring -mstrict-align -meabi       -D_GNU_SOURCE -D_DEFAULT_SOURCE            -DUNIX      -O2 -g -ffunction-sections -fdata-sections   -Wall -Werror-implicit-function-declaration       -DMY_DO_BOOTP=NULL -DHAVE_PPCBUG -DNVRAM_INDIRECT    -D__LINUX_ERRNO_EXTENSIONS__ -DRTEMS_LEGACY_STACK  -I. -I../O.Common -I. -I. -I.. -I../../../../include/compiler/gcc -I../../../../include/os/RTEMS -I../../../../include         -DRTEMS_GZ=libComTestHarness.app.bin.gz -c ../../../../include/os/RTEMS/ppc_bug.S -o libComTestHarness.boot.o
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-ld -o libComTestHarness.boot.elf /gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/bootloader.o --just-symbols=libComTestHarness -b binary libComTestHarness.boot.o --no-warn-mismatch -T /gem_base/targetOS/RTEMS//rtems/5/powerpc-rtems5/mvme2307/lib/ppcboot.lds -Map libComTestHarness.boot.map
/gem_base/targetOS/RTEMS/rtems/5/bin/powerpc-rtems5-ld: cannot find rtems.gz
make[4]: *** [../../../../configure/RULES_BUILD:353: libComTestHarness.boot] Error 1
make[4]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules/libcom/test/O.RTEMS-mvme2700'
make[3]: *** [../../../configure/RULES_ARCHS:58: build.RTEMS-mvme2700] Error 2
make[3]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules/libcom/test'
make[2]: *** [../../configure/RULES_DIRS:85: test.build.RTEMS-mvme2700] Error 2
make[2]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules/libcom'
make[1]: *** [../configure/RULES_DIRS:85: libcom.build.RTEMS-mvme2700] Error 2
make[1]: Leaving directory '/home/mrippa/work/vendor/epics-base/modules'
make: *** [configure/RULES_DIRS:85: modules.build.RTEMS-mvme2700] Error 2


```

mdavidsaver commented:
> ... It always complained about finding rtems.gz ...

Then I can see no reasonable way to allow for parallel make to be safe.  Thank you for humoring me.

anjohnson commented:
I see the problem as being due to our re-using the `ppcboot.lds` linker script from RTEMS, which explicitly mentions the filename rtems.gz:
```
  .image :
  {
    rtems.gz(*)
    . = ALIGN(4);
    *.gz(*)
  }
```
It should be possible to copy and modify that linker script into Base if we could make a more generic linker script, or into the build directory if we have to make it specific to the binary we're building. I think I tried to do that, but had difficulties with setting these variables, which something like your approach might help resolve:
```
    _rtems_gz_size = ABSOLUTE(_binary_rtems_gz_end - _binary_rtems_gz_start);
    _rtems_size = ABSOLUTE(__rtems_end - __rtems_start);
```
Otherwise we have to say that when doing parallel builds EPICS can only build a single munched program in any application Makefile. Base itself currently follows that requirement, I already fixed the places where this was an issue.
